sway/tree/workspace.c

```diff
@@ -513,3 +514,108 @@ struct sway_container *workspace_output_get_highest_available(
 
 	return NULL;
 }
+
+struct pid_workspace {
+	pid_t pid;
+	char *workspace;
+	struct timespec time_added;
+
+	struct sway_container *output;
+	struct wl_listener output_destroy;
+
+	struct wl_list link;
+};
+
+static struct wl_list pid_workspaces;
+
+struct sway_container *workspace_for_pid(pid_t pid) {
+	if (!pid_workspaces.prev && !pid_workspaces.next) {
+		wl_list_init(&pid_workspaces);
+		return NULL;
+	}
+
+	struct sway_container *ws = NULL;
+	struct pid_workspace *pw = NULL;
+
+	wlr_log(L_DEBUG, "Looking up workspace for pid %d", pid);
+
+	do {
+		struct pid_workspace *_pw = NULL;
+		wl_list_for_each(_pw, &pid_workspaces, link) {
+			if (pid == _pw->pid) {
+				wlr_log(L_DEBUG,
+						"found pid_workspace for pid %d, workspace %s",
+						pid, pw->workspace);
+				pw = _pw;
+				goto found;
+			}
+		}
+		pid = get_parent_pid(pid);
+	} while (pid > 0 && pid != getpid());
```

Comment:
I don't think it's worth checking for `pid != getpid()` here, we double-fork so that should never happen (except for swaybar and a few exceptions maybe), and if it does happen we'll still bail out because sway probably isn't too far from pid 1.
We could spare one iteration and stop a pid > 1 though, I don't think init will ever be in a workspace :D